File: java/jzoffer/chapter3/GetKthFromEnd_22.java
package chapter3;

public class GetKthFromEnd_22 {

    public class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
        }
    }

    public ListNode getKthFromEnd(ListNode head, int k) {
        if (head == null) return null;
        ListNode fast = head;
        while (k > 0) {
            if (fast == null) return null;
            fast = fast.next;
            k--;
        }
        ListNode last = head;
        while (fast != null) {
            last = last.next;
            fast = fast.next;
        }
        return last;
    }
}

File: go/jzoffer/getKthFromEnd_22.go
package jzoffer

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func getKthFromEnd(head *ListNode, k int) *ListNode {
	fast, slow := head, head

	for k > 0 {
		if fast == nil {
			return nil
		}
		fast = fast.Next
		k--
	}

	for fast != nil {
		slow = slow.Next
		fast = fast.Next
	}
	return slow
}
